### ISSUE TITLE
implement PGD in PyTorch

### ISSUE THREAD

**Issue by npapernot** [tensorflow/cleverhans#1020]

**Comment by michaelshiyu** [comment]
@npapernot I can take this one. 

I do have a question, though. Since PyTorch is dynamic in the sense that it does not explicitly build a graph before each run, are we ok with the potential inconsistencies in the API if we go for a fully-PyTorch implementation? 

Some inconsistencies I can think of include, for example, when the user uses a PyTorch implementation, the “x” being fed into attack.generate would be the actual data in contrast to a placeholder when the user uses a TF implementation of the same attack.

Thanks.

**Comment by npapernot** [comment]
We thought about this and are not expecting a unified interface across frameworks. 

Instead, we are expecting this kind of API: 

``` 
def pgd(model_fn, input, eps=0.2, iterations=1000):
  return adv_x
```

where `input` and `adv_x` could be numpy arrays. 

cc @david-berthelot who may have some thoughts as well or can correct me if I am not reflecting  properly some of the discussions we had offline

**Comment by iArunava** [comment]
This is a possible implementation of the generate function.
```
    def generate(self, x, **kwargs):
        """
        Generate symbolic graph for adversarial examples and return.
        Args:
            x: The model's symbolic inputs.
            kwargs: See `parse_params`
        """

        assert self.parse_params(**kwargs)

        asserts = []

        # If a data range was specified, check that the input was in that range
        if self.clip_min is not None:
            asserts.append(x.any() >= self.clip_min)

        if self.clip_max is not None:
            asserts.append(x.any() <= self.clip_max)

        # Initialize loop variables
        if self.rand_init:
            eta = torch.FloatTensor(*x.shape).uniform_(-self.minmax, self.minmax)
        else:
            eta = torch.zeros_like(x)

        # Clip eta
        eta = clip_eta(eta, self.ord, self.eps)
        adv_x = x + eta

        if self.clip_min is not None or self.clip_max is not None:
            adv_x = torch.clamp(adv_x, self.clip_max, self.clip_min)

        if self.y_target is not None:
            y = self.y_target
            targeted = True
        elif self.y is not None:
            y = self.y
            targeted = False
        else:
            model_preds = self.model.get_probs(x)
            preds_max = reduce_max(model_preds, 1)
            y = torch.equals(model_preds, preds_max).float()
            y.requires_grad = False
            targeted = False
            del model_preds

        y_kwarg = 'y_target' if targeted else 'y'
        fgm_params = {
            'eps' : self.eps_iter,
            y_kwarg: y,
            'ord' : self.ord,
            'clip_min' : self.clip_min,
            'clip_max' : self.clip_max
        }

        if self.ord == 1:
            raise NotImplementedError("It's not clear that FGM is a good inner loop"
                                      " step for PGD when ord=1, because ord=1 FGM "
                                      " changes only one pixel at a time. We need "
                                      " to rigoursly test a strong ord=1 PGD "
                                      " before enabling this feature.")

        FGM = self.FGM_CLASS(self.model, dtype)

        while i < self.nb_iter:
            """
            Do a projected gradient step.
            """
            adv_x = FGM.generate(adv_x, **fgm_params)

            # Clipping perturbation eta to self.ord norm ball
            eta = adv_x - x
            eta = clip_eta(eta, self.ord, self.eps)
            adv_x = x + eta

            # Redo the clipping.
            # FGM alread already did it, but subtracting and re-adding eta can add some
            # small numerical error
            if self.clip_min is not None or self.clip_max is not None:
                adv_x = torch.clamp(adv_x, self.clip_min, self.clip_max)

        # Asserts run only on CPU
        # When multi-GPU eval code tries to force all PGD ops onto GPU, this
        # can cause an error.
        common_dtype = torch.float32
        # NOTE Maybe this needs a cast
        asserts.append(self.eps <= (1e6 + self.clip_max - self.clip_min))
        
        return adv_x
```

**Comment by iArunava** [comment]
Hi! 
I am currently implementing the attacks in pytorch 1.0 and I was unaware of the of pytorch support on cleverhans, I will be happy to help :)

**Comment by michaelshiyu** [comment]
@iArunava which attacks have you already implemented and would like to add here? I’m asking so that I can take the ones that you haven’t done yet. Thanks.

**Comment by michaelshiyu** [comment]
@npapernot I think what you guys are trying to do is better than having a uniform interface across backends.

So do we want the model_fn to be wrapped as a cleverhans.models.Model? I see the current wrapper is TF-based. I can implement one for PyTorch models if that is needed. Thanks.

**Comment by david-berthelot** [comment]
Actually we were thinking of dropping the model API. The idea is to simplify integration by sampling having a model_func that takes an input and returns the logits.

**Comment by michaelshiyu** [comment]
@david-berthelot Gotcha. Thanks for answering.

**Comment by iArunava** [comment]
Hi :) As I am seeing now, the pytorch models are converted to tf and then the attacks are performed, so where exactly a PyTorch implementation of PGD is going to sit in the library and how its supposed to integrate with the library? Am I missing something?

**Comment by david-berthelot** [comment]
We're dropping the cross-framework approach for the sake of simplicity. The focus is now to make it simple to add an attack to an existing model (say for adversarial training).

The way we see it is as a standalone implementation. So in each framework (TF2, JAX, Torch), there'll be a pgd implementation that follows this API:
```
def pgd(model_fn, input, eps=0.2, iterations=1000):
  # model_fn takes an input batch and produces the corresponding logits.
  # pgd code goes here.
  return adv_input
```

**Comment by michaelshiyu** [comment]
@david-berthelot I'm working on this and I have a quick question. 

@npapernot mentioned that attacks should also support numpy arrays. My idea is to just wrap them as PyTorch objects in the beginning so that inside pgd, everything can be written in pure PyTorch and the code would be cleaner. 

Does this sound good to you? My concern is that when people explicitly feed in numpy objects, they may want to stay with numpy operations throughout to avoid using GPU (or for some other reasons).

Thanks.

**Comment by david-berthelot** [comment]
I'm not sure, as far as I know, TF2 can take numpy arrays directly without conversion. For PyTorch, may be just expecting a Tensor would seem to make sense to me and leave it to the caller to convert their numpy array to a tensor. And I think using the GPU would be expected, the other way around would be a surprise to me (like defaulting to CPU).

@npapernot What's your take on this?

**Comment by michaelshiyu** [comment]
@david-berthelot In PyTorch v1.0, I don't think numpy arrays can always be processed directly without error. For example, PyTorch's Tensor operations would throw a TypeError if a numpy array is fed.

I agree with supporting PyTorch Tensor only since a conversion before calling the function would be quite trivial and harmless for the user.

Thanks.

**Comment by npapernot** [comment]
Completely agree with @david-berthelot

Thanks for the details @michaelshiyu: I'm just getting started with PyTorch so I'm not very familiar with the typical pipeline yet. If it's more inline with the code style established in the PyTorch community, then we should expect a PyTorch tensor

**Comment by michaelshiyu** [comment]
@npapernot Perfect!

I opened PR  #1028 for this issue. Although, since PGD relies on an FGM engine, I think I will implement that as well. Is it ok to do that in the same PR or is it preferable to open a separate PR for FGM?

Thanks.

**Comment by michaelshiyu** [comment]
Also, please ignore the closed PR #1027. I got the branch name wrong when I opened it. Sorry about that.

**Comment by npapernot** [comment]
We had a simpler API in mind (without attack classes), would you mind holding on for a few days so we can clarify our intended API? I took a quick look at your PR and I suspect very few changes would be needed to the code itself but it looks like all we will be keeping is the `generate` method as a standalone function. I just want to reduce the amount of work needed on your end, thanks

**Comment by michaelshiyu** [comment]
@npapernot I see. I will be waiting for the API template. Just let me know when you guys have it ready. Thanks.

**Comment by michaelshiyu** [comment]
@npapernot I made some commits in the PR just to keep track of my work. The current implementations don't conform to the desired function-based API but I will revise everything once you guys have the API template ready.

Thanks.

**Comment by npapernot** [comment]
Thanks! Yes, it should just be a little of refactoring but the underlying code should be the same

**Comment by iArunava** [comment]
Hi, Sorry for the late reply. Since @michaelshiyu is already on PGD I can start taking another attack,or help Michael on that. Let me know.

**Comment by iArunava** [comment]
I created a Pull request for the very simple Noise attack here #1029 
Its WIP. I will make the changes and tag one of you to review. And I am looking forward to the API template will make changes according to that.

**Comment by HaojieYuan** [comment]
Super cool!
I've always expecting better support for pytorch in cleverhans and here it will be! :laughing: 
Would be extremely happy to help :raised_hand_with_fingers_splayed:

**Comment by npapernot** [comment]
Thank you everyone for the contributions, we appreciate your work. @david-berthelot and I have been working on the API offline and are making the changes directly to @michaelshiyu 's PR #1028. We'll iterate a bit more on that PR and close this issue when we are done and have merged the FGM and PGD attacks in PyTorch. At that point it will make sense to include more attacks @HaojieYuan and to adapt and merge your PR @iArunava

**Comment by iArunava** [comment]
Sounds good. I am already implementing `saliency_map_method` attack in PyTorch. Just let me know when you are ready. I will make the PR.

**Comment by michaelshiyu** [comment]
@npapernot awesome! Thanks!

**Comment by npapernot** [comment]
Thanks @iArunava !

**Comment by npapernot** [comment]
@iArunava @michaelshiyu  the attacks were just merged, thanks for all the work! Feel free to open more PRs following a similar API
